first attempt (#61)
oauthstuff/draft-oauth-rar: main.md

The authorization server will consider this audience restriction in the user consent if needed.

In the following example (line breaks for illustrative purposes only), 

```
GET /authorize?response_type=code
   &client_id=s6BhdRkqt3
   &state=af0ifjsldkj
   &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
   &code_challenge_method=S256
   &code_challenge=K2-ltc83acc4h0c9w6ESC_rEMTJ3bwc-uCHaoeK1t8U
   &resource=https%3A%2F%2Fexample%2Ecom%2Fpayments
   &authorization_details=%5B%7B%22type%22%3A%22account%5Finformation%22%2C%22actions
   %22%3A%5B%22list%5Faccounts%22%2C%22read%5Fbalances%22%2C%22read%5Ftransactions%22
   %5D%2C%22locations%22%3A%5B%22https%3A%2F%2Fexample%2Ecom%2Faccounts%22%5D%7D%2C%7
   B%22type%22%3A%22payment%5Finitiation%22%2C%22actions%22%3A%5B%22initiate%22%2C%22
   status%22%2C%22cancel%22%5D%2C%22instructedAmount%22%3A%7B%22currency%22%3A%22EUR%
   22%2C%22amount%22%3A%22123%2E50%22%7D%2C%22creditorName%22%3A%22Merchant123%22%2C%
   2creditorAccount%22%3A%7B%22iban%22%3A%22DE02100100109307118603%22%7D%2C%22remitta

>    nceInformationUnstructured%22%3A%22RefNumberMerchant%22%7D%5D HTTP/1.1

Reviewer: Not sure what's going on here but this seems to decode into some JSON that isn't totally valid.

Author: Good catch! A quotation mark had been replaced by a comma (I don't know why). Fixed

]
```

along with a `resource` parameter value "https://example.com/payments". 

Application of the rules given above will assign the URI "https://example.com/payments" as audience to the object of type "payment_initiation". 

# Authorization Response

This specification does not define extensions to the authorization response. 

# Authorization Error Response

The AS MUST refuse to process any unknown authorization data type or authorization details not conforming to the respective type definition. If any of the objects in `authorization_details` contains an unknown authorization data type or an object of 
known type but containing unknown elements or elements of the wrong type, the AS MUST abort processing and respond with an error `invalid_authorization_details` to the client. 

# Token Request

> The `authorization_details` request parameter MAY be used in a token request to further restrict the permissions assigned to an access token. As an example, the grant approved in the running example above could be restricted as follows

Reviewer: I really dislike this and don't think the value is worth the complexity.

Author: While restructuring it become obvious to me that we did not specify so far whether and how the authorization_details parameter could be used at the token request. We had comments on the list asking for it to downgrade the permissions associated to an access token. I don't have a use case for it and could also live with removing it again. 
@jricher what do you think?

Reviewer: I'm somewhat skeptical of the actual need for downgrade/scope. But if there is a need, I'd like to consider other ways of addressing it that might be more straightforward than application specific constructs that have to determine if one chunk of JSON is a proper down-scoped subset of another. That sounds very hard to implement and a likely vector for accidentally allowing for up-scoping beyond the actual granted authz.

Reviewer 2: I get why it's there, it's in parallel to how `scope` can be used and so that much makes sense. However, `scope` is much, much simpler to describe what "downgrade" means: you remove one of the scope values from the list, probably. With RAR it's a much more complicated calculation for the AS to determine if it's "less" or not. But even then, as long as the AS fails "closed" in cases where it can't tell, and it denies the request, then it should be OK in practice. 

What if the rubric was instead "would not require additional consent from the RO"? Or is that getting to deep into UX territory?

Author: "What if the rubric was instead "would not require additional consent from the RO"? "

What would change then? If I understand it right, the AS would still need to determine automatically whether the authz details match already consented authz details.

Reviewer: I thought that no additional consent was kinda implied by the "to further restrict the permissions assigned" already. 
My push-back on the feature is about the added implementation complexity (with potential for introducing security issues) vs. actual utility of it.

Author: I removed this paragraph. So authorization_details is authz request only now and resource is token request only. Are we good to merge now?

```JSON
[
   {
      "type": "account_information",
      "actions": [
         "list_accounts"
      ],
      "locations": [
         "https://example.com/accounts"
      ]
   }
]
```

to the permission for obtaining the list of accounts. 

The way authorization details in the token request are matched against authorization details attached to the grant being processed is application specific and out of scope for this specification. 

> The `resource` parameter MAY be used in the token request to request the creation of an audience restriected access token. If the client uses this parameter, the AS MUST consider the audience restriction defined by the `locations` elements of the `authorization_details` and the value(s) of `resource` authorization request parameters to filter the authorization data objects applicable to the respective resource(s). 

Reviewer: ```suggestion
The `resource` parameter MAY be used in the token request to request the creation of an audience restricted access token. If the client uses this parameter, the AS MUST consider the audience restriction defined by the `locations` elements of the `authorization_details` and the value(s) of `resource` authorization request parameters to filter the authorization data objects applicable to the respective resource(s). 
```

}
```

TBD: response does not indicate access token audience "https://example.com/payments" (no resource response parameter or locations element)

## Enriched authorization details in Token Response

> The authorization details attached to the access token MAY differ from what the client requests. In addition to the user authorizing less than what the client requested, there are use cases where the authorization server enriches the data in an authorization details object. For example, a client may ask for access to account information but leave the decision about the accounts it will be able to access to the user. The user would select the sub set of accounts she wants the client to entitle to access in the course of the authorization process. In order to allow the client to determine the accounts it is entitled to access, the authorization server will add this information to the respective authorization details object. 

Reviewer 2: Use singular they instead of gendered pronouns, where possible:

```suggestion
The authorization details attached to the access token MAY differ from what the client requests. In addition to the user authorizing less than what the client requested, there are use cases where the authorization server enriches the data in an authorization details object. For example, a client may ask for access to account information but leave the decision about the accounts it will be able to access to the user. The user would select the sub set of accounts they want the client to entitle to access in the course of the authorization process. In order to allow the client to determine the accounts it is entitled to access, the authorization server will add this information to the respective authorization details object. 
```

Author: changed

The registration of new authorization data types with the AS is out of scope of this draft. 

# Scope value "openid" and "claims" parameter

OpenID Connect [@OIDC] specifies the JSON-based `claims` request parameter that can be used to specify the claims a client (acting as OpenID Connect Relying Party) wishes to receive in a fine-grained and privacy preserving way as well as assign those claims to a certain delivery mechanisms, i.e. ID Token or userinfo response. 

The combination of the scope value `openid` and the additional parameter `claims` can be used beside `authorization_details` in the same way as every non-OIDC scope value. 

> Alternatively, there could be an authorization data type for OpenID Connect. (#openid) gives an example of how such an authorization data type could look like.

Reviewer 2: ```suggestion
Alternatively, there could be an authorization data type for OpenID Connect. (#openid) gives an example of what such an authorization data type could look like.
```
